contrib/rust/src/python/pants/contrib/rust/tasks/cargo_build.py

```diff
@@ -0,0 +1,550 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import collections
+import copy
+import functools
+import hashlib
+import json
+import os
+import shutil
+
+from future.utils import PY3
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.build_graph.address import Address
+from pants.invalidation.build_invalidator import CacheKeyGenerator
+from pants.util.dirutil import absolute_symlink, read_file, safe_file_dump, safe_mkdir
+
+from pants.contrib.rust.tasks.cargo_fingerprint_strategy import CargoFingerprintStrategy
+from pants.contrib.rust.tasks.cargo_workspace import Workspace
+from pants.contrib.rust.utils.basic_invocation.args_rules import (args_rules, c_flag_rules,
+                                                                  l_flag_rules)
+from pants.contrib.rust.utils.basic_invocation.env_rules import env_rules
+from pants.contrib.rust.utils.basic_invocation_conversion import \
+  convert_into_pants_invocation as convert_basic_into_pants_invocation
+from pants.contrib.rust.utils.custom_build_invocation_conversion import \
+  convert_into_pants_invocation as convert_custom_build_into_pants_invocation
+from pants.contrib.rust.utils.custom_build_output_parsing import (filter_cargo_statements,
+                                                                  parse_multiple_cargo_statements)
+
+
+class Build(Workspace):
+  _build_script_output_cache = {}
+  _build_index = {}
+  _package_out_dirs = {}
+  _libraries_dir_path = None
+  _build_index_dir_path = None
+
+  @classmethod
+  def implementation_version(cls):
+    return super(Build, cls).implementation_version() + [('Cargo_Build', 1)]
+
+  @classmethod
+  def register_options(cls, register):
+    super(Build, cls).register_options(register)
+    register(
```

Comment:
If these options affect the identity of the output (they probably do), they should be marked `fingerprint=True`.